### ISSUE TITLE
Training command for ImageNet2012 only model

### ISSUE THREAD

**Issue by lengstrom** [google-research/vision_transformer#89]
Hi,

Do you have hyperparameters, or better yet commands, that reproduce the ImageNet2012-only trained model given in the paper?

-Logan

**Comment by lucasliunju** [comment]
I also would like to ask this question. I want to pre-train the vit-b-16 model based on imagenet 2012 model. Currently, the pre-train accuracy (just pre-train, not pre-train + fine-tune) is 72.2%. I'm not sure whether we need to use the parameter of representation_size.

**Comment by andsteing** [comment]
We didn't use this codebase for the pre-training so I unfortunately don't have the hyper parameters ready. But all the necessary information should be available in the paper. If you have specific questions I can double check and confirm here...

**Comment by lucasliunju** [comment]
Thanks for your reply. I have several questions about the pre-train on imagenet2012.

1. I would like to ask the top-1 accuracy of pre-training on imagenet-1k. (just pre-training, not pre-training + fine-tuning). I find the result of pre-training + fine-tuning on imagenet-1k is 77.9.
2. I noticed the code just use 99% percent of training data (input_pipeline.py + 48), so I need to use the whole dataset to pre-train the model?
3. The preprocessing method for training data and test data about pre-training. I noticed the pre-processing method for both training data and testing data on this code is inception_crop (input_pipeline.py + 118). So we still need to use inception_crop for both training data and test data when I try to pre-train the model. How about inception_crop when training and random_crop when testing. 

Currently, the top-1 accuracy about pre-training on VIT-B-16 on imagenet-1k is 73.8%. I'm not sure whether I reproduce the result.

Thank you!

Yong

**Comment by andsteing** [comment]
1. Our final pre-training top-1 accuracy was 74.79% for B/16 (after 300 epochs)
2. Using the whole dataset is the way to go (I don't think though there is a noticeable difference to 99%).
3. We use the same `inception_crop=True` for pre-training, but at `crop=224` resolution. Note that the image is simply resized for testing (i.e. `inception_crop=True` is ignored):

https://github.com/google-research/vision_transformer/blob/a51dbfa231262ffd1e8e52c041324e6de644d99b/vit_jax/input_pipeline.py#L136-L138

**Comment by lucasliunju** [comment]
Thanks for your reply. I would like to ask the value of shuffle_buffer_size and pre_fetch. Maybe I need to increase the value (currently the shuffle_buffer_size is 50_000 and pre_fetch is 2).

Thank you!

Yong

**Comment by andsteing** [comment]
We used `shuffle_buffer_size=256_000` and `prefetch_to_device=2` (although that latter should not make a difference wrt final top-1 accuracy).

The value of `resize` is ignored when using `inception_crop=True`.

**Comment by lucasliunju** [comment]
Thank you! Did you use some other tricks, such as label smoothing since I think my current hyper-parameters are right.

Thank you!

Yong

**Comment by andsteing** [comment]
We used a batch size of 4096, gradient clipping at 1.0, Adam optimizer (0.9, 0.999, wd=0.3), and label smoothing of 0.0001.

**Comment by andsteing** [comment]
We used a batch size of 4096, gradient clipping at 1.0, Adam optimizer (0.9, 0.999, wd=0.3), and label smoothing of 0.0001.

**Comment by lucasliunju** [comment]
Thank you. And attention_drop_out is also 0.1?

Yong

**Comment by andsteing** [comment]
No, `attention_dropout_rate=0.0` but `dropout_rate=0.1` (as in all other [configs]())

**Comment by lucasliunju** [comment]
Thank you. And `representation_size` is 768 or None?

Thank you.

Yong

**Comment by andsteing** [comment]
We used `representation_size=768`.

**Comment by lucasliunju** [comment]
Hi,
sorry for trouble again. 
I think I have set all the hyper-parameters but I still cannot reach the top-1 accuracy 74.79/% (learning from scratch on Imagenet-1k).

1. warmup and base_lr
2. image resolution (224)
3. representation_size (768)
4. sigmoid_logit_cross_entropy
5. initialize [head] [bias] -10
6. optimizer adamw (weight_decay=0.3)
7. gradient_clip 1.0
8. label_smoothing 1e-4

Could you give me some advice if possible

Yong

**Comment by andsteing** [comment]
I assume you also use 300 epochs, 3e-3 learning rate, 10k warmup steps then cosine decay, 4096 batch size?

Our validation accuracy develops as follows (with final value=0.7479):

![i1k_val_acc](https://user-images.githubusercontent.com/16050001/120196252-eb286580-c21f-11eb-88d9-d298ac6d0422.png)

**Comment by lucasliunju** [comment]
Thanks for your reply. Yes, epoch 300, 3e-3 lbase_lr, 10k warm_up, cosine and 4096 batch size. I use multi-host tpu (v3-128) to run the code. Therefore, I need to set batch size as 256 on each host, since there are 16 hosts on v3-128.

1. loss function
`
    def sigmoid_xent(*, logits, labels):
      log_p = jax.nn.log_sigmoid(logits)
      log_not_p = jax.nn.log_sigmoid(-logits)
      nll = -jnp.sum(labels * log_p + (1. - labels) * log_not_p, axis=-1)
      return jnp.mean(nll)
`
2. optimizer
`
opt = flax.optim.Adam(beta1=0.9, beta2=0.999, weight_decay=0.3).create(params)
`
3. initialize bias
`
  model = models.KNOWN_MODELS[args.model]
  VisionTransformer = model.partial(num_classes=dataset_info['num_classes'])
  _, params = VisionTransformer.init_by_shape(
      jax.random.PRNGKey(0),
      # Discard the "num_local_devices" dimension for initialization.
      [(batch['image'].shape[1:], batch['image'].dtype.name)])

  """
  pretrained_path = os.path.join(args.vit_pretrained_dir, f'{args.model}.npz')
  params = checkpoint.load_pretrained(
      pretrained_path=pretrained_path,
      init_params=params,
      model_config=models.CONFIGS[args.model],
      logger=logger)
  """

  params["head"]["bias"] = jnp.full_like(params["head"]["bias"], -10.)
`
4. label smoothing
`
    def loss_fn(params, images, labels):
      with flax.nn.stochastic(update_rng):
        logits = vit_fn(params, images, train=True)
        label_smoothing = 1e-4
        num_classes = 1000
        labels = labels * (1 - label_smoothing) + label_smoothing / num_classes
      return sigmoid_xent(logits=logits, labels=labels)
`
5. gradient clipping
`
    grad_clip_norm = 1.0
    gradients, _ = jax.tree_flatten(g)
    g_l2 = jnp.sqrt(sum([jnp.vdot(p, p) for p in gradients]))
    g_factor = jnp.minimum(1.0, grad_clip_norm / g_l2)
    g = jax.tree_map(lambda p: g_factor * p, g)
`

Yong

**Comment by lucasliunju** [comment]
I'm not sure whether that is related to the method of model initialization since I just delete the code that load the pre-trained model.

Yong

**Comment by andsteing** [comment]
(sorry for the delayed reply)

Since you're using `sigmoid_xent` (and not `softmax_xent`) you would probably want to set `labels = labels * (1 - label_smoothing) + label_smoothing` (i.e. not divide by `num_classes`) in your loss when label smoothing.

Otherwise the code you posted here looks fine.

What kind of performance *do* you currently get? Is it similar to the plot I shared above or completely off?
